### PR TITLE
network: pin file descriptor of persistent storage

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

**Review by bluca — APPROVED** [review]

File: src/network/networkd-manager-varlink.c [hunk 1]
```diff
@@ -185,11 +185,17 @@ static int vl_method_set_persistent_storage(Varlink *vlink, JsonVariant *paramet
                 return r;
 
         if (ready) {
-                r = path_is_read_only_fs("/var/lib/systemd/network/");
+                _cleanup_close_ int fd = -EBADF;
```
Comment (poettering) [inline src/network/networkd-manager-varlink.c:188]: hmm, why do you close the fd again, if you need to reopen it later anyway?

File: src/network/networkd-manager-varlink.c [hunk 2]
```diff
@@ -203,16 +209,16 @@ static int vl_method_set_persistent_storage(Varlink *vlink, JsonVariant *paramet
...
-                r = touch("/run/systemd/netif/persistent-storage-ready");
-                if (r < 0)
-                        log_debug_errno(r, "Failed to create /run/systemd/netif/persistent-storage-ready, ignoring: %m");
-        } else {
-                if (unlink("/run/systemd/netif/persistent-storage-ready") < 0 && errno != ENOENT)
-                        log_debug_errno(errno, "Failed to remove /run/systemd/netif/persistent-storage-ready, ignoring: %m");
-        }
+                _cleanup_close_ int fd = -EBADF;
+
+                fd = open("/var/lib/systemd/network/", O_CLOEXEC | O_DIRECTORY | O_PATH);
```
Comment (poettering) [inline src/network/networkd-manager-varlink.c:228]: here you open it a 2nd time in the same function? Why not just pin it?

Comment (poettering) [reply src/network/networkd-manager-varlink.c:228]: Why O_PATH btw? it makes sense to fail here if the dir has bad access modes for us. Hence I think we should open the dir properly here.

File: src/network/networkd-manager-varlink.c [hunk 3]
```diff
@@ -185,11 +185,17 @@ static int vl_method_set_persistent_storage(Varlink *vlink, JsonVariant *paramet
                 return r;
 
         if (ready) {
-                r = path_is_read_only_fs("/var/lib/systemd/network/");
+                _cleanup_close_ int fd = -EBADF;
+
+                fd = open("/var/lib/systemd/network/", O_CLOEXEC | O_DIRECTORY | O_PATH);
```
Comment (poettering) [inline src/network/networkd-manager-varlink.c:191]: xopenat() with O_DIRECTORY|O_CREAT would be cool here I think.

File: src/network/networkd-manager-varlink.c [hunk 4]
```diff
@@ -198,6 +199,18 @@ static int vl_method_set_persistent_storage(Varlink *vlink, JsonVariant *paramet
                         log_warning("The persistent storage is on read-only filesystem.");
                         return varlink_error(vlink, "io.systemd.Network.StorageReadOnly", NULL);
                 }
+
+                if (fstat(fd, &st) < 0)
+                        return log_warning_errno(r, "Failed to stat the persistent storage: %m");
```
Comment (poettering) [inline src/network/networkd-manager-varlink.c:204]: 2nd arg must be `errno`, not `r`

**Comment by poettering** [comment]
some post-merge comments

File: src/network/networkd-manager-varlink.c (hunk 3 above)
Comment (yuwata) [reply src/network/networkd-manager-varlink.c:191]: networkd cannot create directory under `/var/lib/systemd/`, so `O_CREAT` cannot help us.

**Pull request merged**